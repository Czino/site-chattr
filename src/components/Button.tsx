import { ButtonHTMLAttributes } from 'react'

export const Button = (props: ButtonHTMLAttributes<HTMLButtonElement>) => (
    <button
        {...props}
        className={[
            'bg-gradient-purple text-white p-4 text-xl uppercase font-bold rounded hover:bg-gradient-purple-dark',
            props.className,
        ].join(' ')}
    ></button>
)
