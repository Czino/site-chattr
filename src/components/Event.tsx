import { NDKEvent } from '@nostr-dev-kit/ndk'
import { useIsSelf } from '../hooks/useIsSelf'

const DEFAULTS = {
    USER_NAME: 'Anon',
}
const MS_PER_S = 1000
const GATEWAY = 'https://njump.me/'
export const Event = ({ event }: { event: NDKEvent }) => {
    const isSelf = useIsSelf()

    return (
        <li
            className={[
                'w-full rounded p-4 overflow-hidden flex flex-col gap-1',
                isSelf(event.author.pubkey) ? 'bg-highlight' : 'bg-purple-100',
            ].join(' ')}
        >
            <a className="flex gap-1 items-center leading-5"
                href={GATEWAY + event.author.npub}
                target="_blank"
                rel="noreferrer"
            >
                {event.author.profile?.image ? (
                    <img src={event.author.profile?.image} alt="" className="w-5 h-5 rounded-full" />
                ) : (
                    <div className="w-5 h-5 rounded-full bg-purple-400"></div>
                )}
                <p className="font-bold">{event.author.profile?.displayName || DEFAULTS.USER_NAME}</p>
                {!!event.created_at && (
                    <p className="text-purple-400 text-xs">
                        {new Date(event.created_at * MS_PER_S).toLocaleString()}
                    </p>
                )}
            </a>
            <p>{event.content}</p>
        </li>
    )
}
