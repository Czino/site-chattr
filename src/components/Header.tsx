import { getDisplayUrl } from '../helpers/getDisplayUrl'

type Props = { url: string }
export const Header = ({ url }: Props) => (
    <header className="sticky top-0 p-2 text-center bg-purple-400">
        <h1 className="text-purple-50">SiteChattr</h1>
        <h2 className="overflow-x-auto text-purple-50">{getDisplayUrl(url)}</h2>
    </header>
)
