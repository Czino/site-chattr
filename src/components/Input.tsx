import { ChangeEvent, InputHTMLAttributes } from 'react'

type Props = InputHTMLAttributes<HTMLInputElement> & {
    onTextChange?: (value: string) => void
}
export const Input = ({ onTextChange, ...props }: Props) => {
    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        if (onTextChange) onTextChange(event.target.value)
    }
    return (
        <input
            {...props}
            onChange={handleChange}
            className={[
                'resize-none border border-gray-300 rounded-md p-2 flex-shrink-0',
                'focus:outline-none focus:ring-2 focus:ring-blue-500',
            ].join(' ')}
        />
    )
}
