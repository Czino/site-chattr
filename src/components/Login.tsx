import { NDKPrivateKeySigner } from '@nostr-dev-kit/ndk'
import { atom, useAtom } from 'jotai'
import { useActiveUser, useLogin } from 'nostr-hooks'
import { useEffect } from 'react'

export const isLoggedInAtom = atom(false)

export const Login = () => {
    const [, setIsLoggedIn] = useAtom(isLoggedInAtom)
    const { loginFromLocalStorage, loginWithExtension, loginWithSecretKey } = useLogin()
    const createNewPrivateKey = () => {
        const newSigner = NDKPrivateKeySigner.generate()
        return newSigner.privateKey
    }
    const onError = (e: unknown) => {
        throw new Error(String(e))
    }
    const { activeUser } = useActiveUser()

    useEffect(() => loginFromLocalStorage(), [loginFromLocalStorage])
    useEffect(() => {
        setIsLoggedIn(!!activeUser)
    }, [activeUser, setIsLoggedIn])

    return (
        <div className="grid grid-cols-1 gap-4">
            <button onClick={() => loginWithExtension({ onError })}>Login with Extension</button>
            {/* <button onClick={() => loginWithRemoteSigner({ onError })}>Login with Remote Signer</button> */}
            {/* <button
                onClick={() => loginWithSecretKey({ secretKey: prompt('Paste your secret here') || '', onError })}>
                Login with Secret Key
            </button> */}
            <button onClick={() => loginWithSecretKey({ secretKey: createNewPrivateKey(), onError })}>
                Create new user
            </button>
        </div>
    )
}
