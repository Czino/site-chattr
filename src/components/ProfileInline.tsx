import { NDKUserProfile } from '@nostr-dev-kit/ndk'

const DEFAULTS = {
    USER_NAME: 'Anon',
    AVATAR: require('../assets/default_avatar.png'),
}
export const ProfileInline = ({ profile }: { profile: NDKUserProfile }) => (
    <>
        <img src={profile.image || DEFAULTS.AVATAR} alt="" className="w-5 h-5 bg-purple-50 rounded-full" />
        <p className="font-bold">{profile.displayName || DEFAULTS.USER_NAME}</p>
    </>
)
