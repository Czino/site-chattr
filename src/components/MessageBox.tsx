import { useState } from 'react'
import { IoMdSend } from 'react-icons/io'
import { usePostMessage } from '../hooks/usePostMessage'
import { Button } from './Button'
import { Textarea } from './Textarea'

type Props = { url: string }
export const MessageBox = ({ url }: Props) => {
    const [content, setContent] = useState('')
    const clearContent = () => setContent('')
    const postMessage = usePostMessage({
        content,
        url,
        onSuccess: clearContent,
    })

    return (
        <div className="flex flex-col gap-2">
            <div className="flex flex-row flex-nowrap gap-2 items-end p-2">
                <Textarea
                    className="flex-shrink pr-12 w-full"
                    id="textarea"
                    value={content}
                    onInput={(e) => {
                        const target = e.target as HTMLTextAreaElement
                        target.style.height = 'auto'
                        target.style.height = `${target.scrollHeight}px`
                    }}
                    onTextChange={setContent}
                    placeholder="Your message..."
                />
                <Button
                    onClick={postMessage}
                    aria-label="Send"
                    className="flex absolute right-3 bottom-3 justify-center items-center p-0 w-10 h-10 rounded-full"
                >
                    <IoMdSend className="w-7 h-7 relative left-[2px] flex-shrink-0" />
                </Button>
            </div>
        </div>
    )
}
