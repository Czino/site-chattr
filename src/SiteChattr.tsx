import { useAtom } from 'jotai'
import { useNostrHooks } from 'nostr-hooks'
import { Events } from './Events'
import { Header } from './components/Header'
import { isLoggedInAtom, Login } from './components/Login'
import { MessageBox } from './components/MessageBox'
import { urlTypeAtom, URLTypeSelector } from './components/URLTypeSelector'
import { getURLOrigin } from './helpers/getURLOrigin'
import { useCurrentUrl } from './hooks/useCurrentUrl'
import { useCustomNdk } from './hooks/useCustomNdk'

export const SiteChattr = () => {
    const [currentUrl] = useCurrentUrl()
    const [customNdk] = useCustomNdk()
    const [type] = useAtom(urlTypeAtom)

    useNostrHooks(customNdk)
    const [isLoggedIn] = useAtom(isLoggedInAtom)

    return (
        <div className="flex flex-col gap-4 h-full">
            <main className="flex flex-col flex-shrink h-full">
                <Header url={currentUrl} />
                <URLTypeSelector />
                <div className="overflow-auto flex-shrink p-2 h-full">
                    <Events url={type === 'page' ? currentUrl : getURLOrigin(currentUrl)} />
                </div>
                {isLoggedIn ? (
                    <div className="px-2 pb-2">
                        <MessageBox url={currentUrl} />
                    </div>
                ) : (
                    <Login />
                )}
            </main>
        </div>
    )
}
