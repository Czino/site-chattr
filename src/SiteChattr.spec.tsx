import { render, renderHook } from '@testing-library/react'
import { useAtom } from 'jotai'
import { SiteChattr } from './SiteChattr'
import { isLoggedInAtom } from './components/Login'

describe('SiteChattr', () => {
    const base = render(<SiteChattr />).asFragment()
    it('should render correctly when logged in', () => {
        expect(base).toMatchSnapshot()
    })
    it('should render correctly when not logged in', () => {
        const { asFragment } = render(<SiteChattr />)
        renderHook(() => {
            const [, setIsLoggedIn] = useAtom(isLoggedInAtom)
            setIsLoggedIn(false)
        })
        expect(asFragment()).toMatchDiffSnapshot(base)
    })
})
