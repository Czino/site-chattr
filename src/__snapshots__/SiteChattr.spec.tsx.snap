// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteChattr should render correctly when logged in 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 h-full"
  >
    <main
      class="flex flex-col flex-shrink h-full"
    >
      <header
        class="sticky top-0 p-2 text-center bg-purple-400"
      >
        <h1
          class="text-purple-50"
        >
          SiteChattr
        </h1>
        <h2
          class="overflow-x-auto text-purple-50"
        >
          localhost/
        </h2>
      </header>
      <div
        class="flex flex-row"
      >
        <button
          class="relative p-4 w-full font-bold text-center uppercase bg-purple-50"
        >
          Page
          <div
            class="absolute right-0 bottom-0 left-0 w-full h-1 bg-highlight"
          />
        </button>
        <button
          class="relative p-4 w-full font-bold text-center uppercase bg-purple-50"
        >
          Site
        </button>
      </div>
      <div
        class="overflow-auto flex-shrink p-2 h-full"
      >
        <div
          class="text-center text-purple-300"
        >
          Fetching messages...
        </div>
      </div>
      <div
        class="grid grid-cols-1 gap-2 text-center"
      >
        <button
          class="bg-gradient-purple text-white p-4 text-xl uppercase font-bold rounded hover:bg-gradient-purple-dark "
        >
          Login
        </button>
      </div>
    </main>
  </div>
</DocumentFragment>
`;

exports[`SiteChattr should render correctly when not logged in 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
